# Módulo 07, Tarefa 01

Vamos começar a mexer na nossa base de projeto? Já fizemos um exercício de montar a variável resposta, já pudemos perceber que essa atividade pode não ser trivial. Vamos agora trabalhar a base para que fique propícia ao *scikitlearn* para trabalharmos.

Lembrando, a base se chama demo01.csv, e originalmente está publicada [aqui](https://www.kaggle.com/rikdifos/credit-card-approval-prediction).

#### 1) Carregue a base e avalie:

- As variáveis
- Tipos de dados de cada variável
- Quantidade de missings
- Distribuição da variável resposta (mau)

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns

In [ ]:
df = pd.read_csv("demo01.csv")
df.head(1)

sexo posse_de_veiculo posse_de_imovel  qtd_filhos tipo_renda  \
0    M                Y               Y           0    Working   

                        educacao estado_civil    tipo_residencia      idade  \
0  Secondary / secondary special      Married  House / apartment  58.832877   

   tempo_emprego  possui_celular  possui_fone_comercial  possui_fone  \
0       3.106849               1                      0            0   

   possui_email  qt_pessoas_residencia    mau  
0             0                    2.0  False

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16650 entries, 0 to 16649
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   sexo                   16650 non-null  object 
 1   posse_de_veiculo       16650 non-null  object 
 2   posse_de_imovel        16650 non-null  object 
 3   qtd_filhos             16650 non-null  int64  
 4   tipo_renda             16650 non-null  object 
 5   educacao               16650 non-null  object 
 6   estado_civil           16650 non-null  object 
 7   tipo_residencia        16650 non-null  object 
 8   idade                  16650 non-null  float64
 9   tempo_emprego          16650 non-null  float64
 10  possui_celular         16650 non-null  int64  
 11  possui_fone_comercial  16650 non-null  int64  
 12  possui_fone            16650 non-null  int64  
 13  possui_email           16650 non-null  int64  
 14  qt_pessoas_residencia  16650 non-null  float64
 15  ma

In [ ]:
df.isna().sum()

sexo                     0
posse_de_veiculo         0
posse_de_imovel          0
qtd_filhos               0
tipo_renda               0
educacao                 0
estado_civil             0
tipo_residencia          0
idade                    0
tempo_emprego            0
possui_celular           0
possui_fone_comercial    0
possui_fone              0
possui_email             0
qt_pessoas_residencia    0
mau                      0
dtype: int64

#### 2) Vamos montar um metadados

1. Crie um dataframe com os nomes de cada variável e o tipo de dados de cada variável.
2. Adicione uma coluna nesse *dataframe* chamada "qtd_categorias" e coloque nela o número de categorias correspondente de cada variável.
    Dica:
        1. inicie uma lista vazia
        2. faça um for ao longo dos nomes das variáveis,
        3. conte o número de categorias dessa variável
        4. acumule essa informação de 3. na lista que você criou em 1.
        5. No final, essa lista pode ser atribuída à nossa variável.

In [ ]:
df2 = pd.DataFrame({
    'vaiaveis': df.columns,
    'tipo': [df[col].dtype for col in df.columns]
})

In [ ]:
categorias = []
for col in df.columns:
    categorias.append(df[col].nunique())

In [ ]:
df2['qtd_categorias'] = categorias

In [ ]:
df2

vaiaveis     tipo  qtd_categorias
0                    sexo   object               2
1        posse_de_veiculo   object               2
2         posse_de_imovel   object               2
3              qtd_filhos    int64               8
4              tipo_renda   object               5
5                educacao   object               5
6            estado_civil   object               5
7         tipo_residencia   object               6
8                   idade  float64            5298
9           tempo_emprego  float64            3005
10         possui_celular    int64               1
11  possui_fone_comercial    int64               2
12            possui_fone    int64               2
13           possui_email    int64               2
14  qt_pessoas_residencia  float64               9
15                    mau     bool               2

#### 3) Crie variáveis dummy para as variáveis necessárias (i.e. aquelas que são qualitativas e não estão armazenadas como {0, 1} ou {True, False}. Crie um *dataframe* apenas com as variáveis apropriadas para entrada no scikitlearn - elimine as variáveis tipo *str*, mantendo apenas suas versões *dummy*.

In [ ]:
quali = df.select_dtypes(include=['object']).columns.tolist()
quali

['sexo',
 'posse_de_veiculo',
 'posse_de_imovel',
 'tipo_renda',
 'educacao',
 'estado_civil',
 'tipo_residencia']

In [ ]:
df_dummies = pd.get_dummies(df, columns=quali, drop_first=True)
df_dummies.head()

qtd_filhos      idade  tempo_emprego  possui_celular  \
0           0  58.832877       3.106849               1   
1           0  52.356164       8.358904               1   
2           0  52.356164       8.358904               1   
3           0  46.224658       2.106849               1   
4           0  29.230137       3.021918               1   

   possui_fone_comercial  possui_fone  possui_email  qt_pessoas_residencia  \
0                      0            0             0                    2.0   
1                      0            1             1                    1.0   
2                      0            1             1                    1.0   
3                      1            1             1                    2.0   
4                      0            0             0                    2.0   

     mau  sexo_M  ...  educacao_Secondary / secondary special  \
0  False    True  ...                                    True   
1  False   False  ...                                    True   
2  False   False  ...                                    True   
3  False    True  ...                                   False   
4  False   False  ...                                   False   

   estado_civil_Married  estado_civil_Separated  \
0                  True                   False   
1                 False                   False   
2                 False                   False   
3                  True                   False   
4                  True                   False   

   estado_civil_Single / not married  estado_civil_Widow  \
0                              False               False   
1                               True               False   
2                               True               False   
3                              False               False   
4                              False               False   

   tipo_residencia_House / apartment  tipo_residencia_Municipal apartment  \
0                               True                                False   
1                               True                                False   
2                               True                                False   
3                               True                                False   
4                               True                                False   

   tipo_residencia_Office apartment  tipo_residencia_Rented apartment  \
0                             False                             False   
1                             False                             False   
2                             False                             False   
3                             False                             False   
4                             False                             False   

   tipo_residencia_With parents  
0                         False  
1                         False  
2                         False  
3                         False  
4                         False  

[5 rows x 29 columns]

In [ ]:
df_dummies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16650 entries, 0 to 16649
Data columns (total 29 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   qtd_filhos                              16650 non-null  int64  
 1   idade                                   16650 non-null  float64
 2   tempo_emprego                           16650 non-null  float64
 3   possui_celular                          16650 non-null  int64  
 4   possui_fone_comercial                   16650 non-null  int64  
 5   possui_fone                             16650 non-null  int64  
 6   possui_email                            16650 non-null  int64  
 7   qt_pessoas_residencia                   16650 non-null  float64
 8   mau                                     16650 non-null  bool   
 9   sexo_M                                  16650 non-null  bool   
 10  posse_de_veiculo_Y                      16650 non-null  bo

In [ ]:
boleanos = df_dummies.select_dtypes(include=['bool']).columns

for a in boleanos:
    df_dummies[a] = df_dummies[a].map({True: 1, False: 0})
df_dummies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16650 entries, 0 to 16649
Data columns (total 29 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   qtd_filhos                              16650 non-null  int64  
 1   idade                                   16650 non-null  float64
 2   tempo_emprego                           16650 non-null  float64
 3   possui_celular                          16650 non-null  int64  
 4   possui_fone_comercial                   16650 non-null  int64  
 5   possui_fone                             16650 non-null  int64  
 6   possui_email                            16650 non-null  int64  
 7   qt_pessoas_residencia                   16650 non-null  float64
 8   mau                                     16650 non-null  int64  
 9   sexo_M                                  16650 non-null  int64  
 10  posse_de_veiculo_Y                      16650 non-null  in

In [ ]:
df_dummies.head()

qtd_filhos      idade  tempo_emprego  possui_celular  \
0           0  58.832877       3.106849               1   
1           0  52.356164       8.358904               1   
2           0  52.356164       8.358904               1   
3           0  46.224658       2.106849               1   
4           0  29.230137       3.021918               1   

   possui_fone_comercial  possui_fone  possui_email  qt_pessoas_residencia  \
0                      0            0             0                    2.0   
1                      0            1             1                    1.0   
2                      0            1             1                    1.0   
3                      1            1             1                    2.0   
4                      0            0             0                    2.0   

   mau  sexo_M  ...  educacao_Secondary / secondary special  \
0    0       1  ...                                       1   
1    0       0  ...                                       1   
2    0       0  ...                                       1   
3    0       1  ...                                       0   
4    0       0  ...                                       0   

   estado_civil_Married  estado_civil_Separated  \
0                     1                       0   
1                     0                       0   
2                     0                       0   
3                     1                       0   
4                     1                       0   

   estado_civil_Single / not married  estado_civil_Widow  \
0                                  0                   0   
1                                  1                   0   
2                                  1                   0   
3                                  0                   0   
4                                  0                   0   

   tipo_residencia_House / apartment  tipo_residencia_Municipal apartment  \
0                                  1                                    0   
1                                  1                                    0   
2                                  1                                    0   
3                                  1                                    0   
4                                  1                                    0   

   tipo_residencia_Office apartment  tipo_residencia_Rented apartment  \
0                                 0                                 0   
1                                 0                                 0   
2                                 0                                 0   
3                                 0                                 0   
4                                 0                                 0   

   tipo_residencia_With parents  
0                             0  
1                             0  
2                             0  
3                             0  
4                             0  

[5 rows x 29 columns]

#### 4) Qual variável é mais poderosa?

Considere as variáveis ```possui_email``` e ```posse_de_veiculo```. Faça uma tabela cruzada entre elas e responda qual delas te parece mais poderosa para prever a probabilidade de ```mau = 1```?

In [ ]:
cross_tab_email = pd.crosstab(df_dummies['possui_email'], df_dummies['mau'], normalize='index')
cross_tab_veiculo = pd.crosstab(df_dummies['posse_de_veiculo_Y'], df_dummies['mau'], normalize='index')


In [ ]:
cross_tab_email

mau                  0         1
possui_email                    
0             0.976269  0.023731
1             0.979730  0.020270

In [ ]:
cross_tab_veiculo

mau                        0         1
posse_de_veiculo_Y                    
0                   0.975732  0.024268
1                   0.977905  0.022095

Ambas as variáveis têm uma distribuição semelhante em relação à probabilidade de mau. No entanto, a variável possui_email apresenta uma ligeira diferença maior entre as categorias (0 e 1) em comparação com posse_de_veiculo. Portanto, possui_email pode ser ligeiramente mais poderosa para prever a probabilidade de mau = 1.

* Mas se for aproundar um pouco mais da pra tentar olhar pelo p valor dos qui-quadrado

In [ ]:
import scipy.stats as stats

# Perform chi-square test for independence for 'possui_email' and 'mau'
chi2_email, p_email, dof_email, ex_email = stats.chi2_contingency(pd.crosstab(df_dummies['possui_email'], df_dummies['mau']))

# Perform chi-square test for independence for 'posse_de_veiculo' and 'mau'
chi2_veiculo, p_veiculo, dof_veiculo, ex_veiculo = stats.chi2_contingency(pd.crosstab(df_dummies['posse_de_veiculo_Y'], df_dummies['mau']))

p_email, p_veiculo

(0.45311674875502095, 0.3947252595418437)

* **P-valores dos testes qui-quadrado:**

possui_email:
p=0.453
p=0.453
posse_de_veiculo:
p=0.395
p=0.395

Os p-valores para ambas as variáveis são maiores que 0.05, indicando que não há evidência estatística suficiente para afirmar que há uma associação significativa entre essas variáveis e a probabilidade de mau = 1.

Com base nos p-valores, nenhuma das variáveis parece ser uma preditora poderosa da probabilidade de mau = 1.


#### 5) Salve a base, pois ela será utilizada no final deste módulo.

In [ ]:
df_dummies.to_csv('demo01_damificado.csv', index=False)

In [ ]:
pd.read_csv('demo01_damificado.csv').head()

qtd_filhos      idade  tempo_emprego  possui_celular  \
0           0  58.832877       3.106849               1   
1           0  52.356164       8.358904               1   
2           0  52.356164       8.358904               1   
3           0  46.224658       2.106849               1   
4           0  29.230137       3.021918               1   

   possui_fone_comercial  possui_fone  possui_email  qt_pessoas_residencia  \
0                      0            0             0                    2.0   
1                      0            1             1                    1.0   
2                      0            1             1                    1.0   
3                      1            1             1                    2.0   
4                      0            0             0                    2.0   

   mau  sexo_M  ...  educacao_Secondary / secondary special  \
0    0       1  ...                                       1   
1    0       0  ...                                       1   
2    0       0  ...                                       1   
3    0       1  ...                                       0   
4    0       0  ...                                       0   

   estado_civil_Married  estado_civil_Separated  \
0                     1                       0   
1                     0                       0   
2                     0                       0   
3                     1                       0   
4                     1                       0   

   estado_civil_Single / not married  estado_civil_Widow  \
0                                  0                   0   
1                                  1                   0   
2                                  1                   0   
3                                  0                   0   
4                                  0                   0   

   tipo_residencia_House / apartment  tipo_residencia_Municipal apartment  \
0                                  1                                    0   
1                                  1                                    0   
2                                  1                                    0   
3                                  1                                    0   
4                                  1                                    0   

   tipo_residencia_Office apartment  tipo_residencia_Rented apartment  \
0                                 0                                 0   
1                                 0                                 0   
2                                 0                                 0   
3                                 0                                 0   
4                                 0                                 0   

   tipo_residencia_With parents  
0                             0  
1                             0  
2                             0  
3                             0  
4                             0  

[5 rows x 29 columns]